File: leet-code/377.组合总和-ⅳ.js
/*
 * @lc app=leetcode.cn id=377 lang=javascript
 *
 * [377] 组合总和 Ⅳ
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var combinationSum4 = function (nums, target) {
  // 完全背包之排列
  const dp = Array(target + 1).fill(0)
  dp[0] = 1

  // 1. 遍历背包
  for (let i = 0; i <= target; i++) {
    // 遍历物品
    for (let j = 0; j < nums.length; j++) {
      const num = nums[j];
      dp[i] += i - num >= 0 ? dp[i - num] : 0
    }
  }
  return dp[target]
};
// @lc code=end


// @after-stub-for-debug-begin
module.exports = combinationSum4;
// @after-stub-for-debug-end

File: leet-code-rs/src/solutions/377_组合总和-ⅳ.rs
/*
 * @lc app=leetcode.cn id=377 lang=rust
 *
 * [377] 组合总和 Ⅳ
 */

// @lc code=start

#[allow(unused)]
impl Solution {
    /**
     * 解题思路：
     *  完全背包求解最大排列数（非组合）
     */
    pub fn combination_sum4(nums: Vec<i32>, target: i32) -> i32 {
        let mut dp = vec![0; target as usize + 1];
        dp[0] = 1;

        // 遍历顺序，因为是求排列，所以要先遍历背包，确保物品顺序被打乱，同时因为是完全背包，所以要从小到大遍历。
        for j in 0..=target as usize {
            // 后遍历物品
            for i in 0..nums.len() {
                if j >= nums[i] as usize {
                    dp[j] += dp[j - nums[i] as usize];
                }
            }
        }

        return dp[target as usize];
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_combination_sum4() {
        //assert_eq!(Solution::combination_sum4(vec![]),[]);
        assert!(true)
    }
}
